File: app/src/main/java/com/github/ykiselev/lwjgl3/fs/ClassPathResources.java
package com.github.ykiselev.lwjgl3.fs;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URL;
import java.util.Objects;
import java.util.Optional;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.Objects.requireNonNull;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public final class ClassPathResources implements ResourceFolder {

    private final ClassLoader loader;

    public ClassPathResources(ClassLoader loader) {
        this.loader = requireNonNull(loader);
    }

    @Override
    public Optional<URL> resolve(String resource, boolean shouldExist) {
        return Stream.of(resource)
                .map(loader::getResource)
                .filter(Objects::nonNull)
                .findFirst();
    }

    @Override
    public Stream<URL> resolveAll(String resource) {
        if (resource == null) {
            return Stream.empty();
        }
        return all(resource);
    }

    private Stream<URL> all(String name) {
        try {
            return StreamSupport.stream(
                    Spliterators.spliteratorUnknownSize(
                            loader.getResources(name).asIterator(),
                            Spliterator.ORDERED
                    ),
                    false
            );
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: app/src/main/kotlin/com/github/ykiselev/playground/services/fs/ClassPathResources.kt
package com.github.ykiselev.playground.services.fs

import java.net.URL

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
class ClassPathResources(private val loader: ClassLoader) : ResourceFolder {

    override fun resolve(resource: String, shouldExist: Boolean): URL? =
        sequenceOf(resource)
            .mapNotNull { name -> loader.getResource(name) }
            .firstOrNull()


    override fun resolveAll(resource: String): Sequence<URL> =
        loader.getResources(resource).asSequence()

}